python/ray/experimental/workflow/common.py

```diff
@@ -0,0 +1,74 @@
+from typing import Tuple, List, Optional, Callable
+import uuid
+
+import ray
+
+# Alias types
+RRef = ray.ObjectRef  # Alias ObjectRef because it is too long in type hints.
+StepID = str
+WorkflowOutputType = RRef
+WorkflowInputTuple = Tuple[RRef, List["Workflow"], List[RRef]]
+StepExecutionFunction = Callable[[StepID, WorkflowInputTuple],
+                                 WorkflowOutputType]
+SerializedStepFunction = str
+
+
+class Workflow:
+    def __init__(self, original_function: Callable,
+                 step_execution_function: StepExecutionFunction,
+                 input_placeholder: RRef, input_workflows: List["Workflow"],
+                 input_object_refs: List[RRef]):
+        self._input_placeholder: RRef = input_placeholder
+        self._input_workflows: List[Workflow] = input_workflows
+        self._input_object_refs: List[RRef] = input_object_refs
+        # we need the original function for checkpointing
+        self._original_function: Callable = original_function
+        self._step_execution_function: StepExecutionFunction = (
+            step_execution_function)
+
+        self._executed: bool = False
+        self._output: Optional[WorkflowOutputType] = None
+        self._step_id: StepID = uuid.uuid4().hex
+
+    @property
+    def executed(self) -> bool:
+        return self._executed
+
+    @property
+    def output(self) -> WorkflowOutputType:
+        if not self._executed:
+            raise Exception("The workflow has not been executed.")
+        return self._output
+
+    @property
+    def id(self) -> StepID:
+        return self._step_id
+
+    def execute(self) -> RRef:
+        """
+        Trigger workflow execution recursively.
+        """
+        if self.executed:
+            return self._output
+        workflow_outputs = [w.execute() for w in self._input_workflows]
```

Comment:
Maybe you have answered it in the meeting, but could you confirm here these workflows will be executed in parallel?

Reply:
yes, they will be executed in parallel